Changes to make transition to wxWidgets3.0 on Windows easier

turtleli opened the pull request:
These commits are to prepare for wxWidgets 3.0 on Windows. Ideally I'm thinking of getting the code to a state where it can be smoothly switched over to wxWidgets 3.0 without breaking anything, but there's still a few issues to be sorted out.

1st patch removes the use of an custom allocator that's been removed from wxWidgets.
2nd patch fixes header file includes (conflict between wxwidgets and w32pthreads, <algorithm> for min and max).
3rd patch adds a function to replace wxCharCodeMSWToWX which was removed. The added function is basically a copy of the Linux one but modified for Windows.
4th patch fixes the command line help message box in Windows (An alternative would be to rip out the customised message box and just use the standard message box for the command line help, it might cut down future maintenance).

Please review.

Things that I know are currently broken in Windows PCSX2 when compiled against wxWidgets 3.0:

PNG loading (Debug build assertions, minor issue) -  libpng included with wxWidgets has upgraded from 1.2 to 1.6, which is a lot stricter (the call stack when the assert is reached suggests iCCP: known incorrect sRGB profile). Stripping/replacing the profiles from the PNGs would probably fix this issue, but I've left it alone since it's actually useful for debugging the main issue.

Strings (All builds, major issue) - See picture. The problem seems to be a combination of:
 * On Windows, wxString.c_str() is a wide string in 2.8, but in wxWidgets 3.0, wxString.c_str() can be a narrow or wide string.
 * wxVsnprintf, which resolves to _vswprintf_p on Windows, treat %s as a wide string but is sometimes passed narrow strings and therefore can't detect a terminating NULL.
 * Lots of conversions to and from wide/narrow strings centred around the FastFormatUnicode and FastFormatAscii classes.

Very open to suggestions on quickly/efficiently fixing the string problem. I've briefly tried changing the definition of WX_STR on Windows, but seems like there's no quick fix for it.
![stringfail](https://cloud.githubusercontent.com/assets/10976277/8093311/08b01650-0fb8-11e5-9c6f-d6e7ba63c0a7.png)

gregory38 commented:
```
Strings (All builds, major issue) - See picture. The problem seems to be a combination of:

    On Windows, wxString.c_str() is a wide string in 2.8, but in wxWidgets 3.0, wxString.c_str() can be a narrow or wide string.
    wxVsnprintf, which resolves to _vswprintf_p on Windows, treat %s as a wide string but is sometimes passed narrow strings and therefore can't detect a terminating NULL.
    Lots of conversions to and from wide/narrow strings centred around the FastFormatUnicode and FastFormatAscii classes.

Very open to suggestions on quickly/efficiently fixing the string problem. I've briefly tried changing the definition of WX_STR on Windows, but seems like there's no quick fix for it.
```
The solution needs to work on linux WX2.8 and linux WX3.0. I think it is the biggest difficulty.

What did you try for the WX_STR macro?

turtleli commented:
I ifdef'd it so it only affects Windows, WX3.0 `#if wxMAJOR_VERSION >= 3`, `#ifdef _WIN32`. I can test on WX3.0 Linux but haven't got a setup for WX2.8 Linux. I've tried the following:
```
#define WX_STR(str) (static_cast<const char_t*>(str.c_str()));
#define WX_STR(str) (static_cast<const wchar_t*>(str.c_str()));
#define WX_STR(str) (static_cast<const wchar_t*>(str.wc_str()));
```
But second and third are pretty much the same.

turtleli commented:
Oh, forgot to mention, `WriteLn( "narrow format string", ...)` goes through vsnprintf, `WriteLn( "wide format string", ...)` goes through wxVsnprintf. But then what approach is better for unicode on both Windows and Linux since one is UTF16 and the other is UTF8?

gregory38 commented:
Well, UTF management is quite hackish and painful. Normally they're tons of cast and conversion. Globally most code is converted to UTF16.

The real issue is the "%s" vs "%ls" vs L"%s" vs L"%ls". Could you give an example of one of the string issue. Did you try mb_str()?

For reference,
```
http://docs.wxwidgets.org/trunk/overview_changes_since28.html
http://docs.wxwidgets.org/trunk/overview_unicode.html
```

ssakash commented:
I think it would be ideal to make a separate branch for this preparation, there are lots of stuffs to be changed if we're changing to wxWidgets3.0

gregory38 commented:
50% of the Linux user base already use wx3.0. So no, except this UTF non-sense, change will remains small and manageable.

turtleli commented:
Yeah, that's the main issue. I've just tried mb_str, with and without the static cast plus fixups, and it also has runtime errors. I suspect there's no quick fix and I might just have to go through the code and make every format string consistent.

My tests on Windows indicate
_vswprintf_p - L"%s" and L"%ls" both expect wide string.
vsnprintf - "%s" truncates wide strings (expected) and "%ls" works as expected.

Example from pcsx2/System.cpp:
```
	Console.Indent().WriteLn(
		L"CPU name         =  %s\n"
		L"Vendor/Model     =  %s (stepping %02X)\n"
		L"CPU speed        =  %u.%03u ghz (%u logical thread%ls)\n"
		L"x86PType         =  %s\n"
		L"x86Flags         =  %08x %08x\n"
		L"x86EFlags        =  %08x",
			WX_STR(fromUTF8( x86caps.FamilyName ).Trim().Trim(false)),
			WX_STR(fromUTF8( x86caps.VendorName )), x86caps.StepID,
			speed / 1000, speed % 1000,
			x86caps.LogicalCores, (x86caps.LogicalCores==1) ? L"" : L"s",
			WX_STR(x86caps.GetTypeName()),
			x86caps.Flags, x86caps.Flags2,
			x86caps.EFlags
	);
```
A cast to char * or mb_str() for WX_STR would output garbage. Choosing wc_str() for WX_STR would work here, but would break the "%s" case in other parts of the code.

turtleli commented:
Made some additional changes
 * Added an extra include (folded into the fix includes commit)
 * Used wc_str for some Windows only code to resolve some ambiguity when compiling with wxWidgets 3.0. (pcsx2/windows/samplprof.cpp seems unused?)
 * Removed wxadv28\*.dll, wxbase28\*dll and wxcore28\*.dll from the additonal linker dependencies section of a property sheet. The libraries are already linked due to the project references property for the SPU2-X and pcsx2 projects.

turtleli commented:
More changes:
 * Include additional linker dependencies for pcsx2 VS2013 project.
 * Fix strings on Windows wxWidgets 3.0. (This also sets WX_STR to wc_str for Windows WX2.8 builds).
 * Fix some logging in wxWidgets 3.0. DoLogRecord was not overridden in pxLogConsole.

Things that I know are currently broken in wx3.0 Windows (I'm using my branch at http://github.com/turtleli/PCSX2/tree/wxwidgets3_revised):
 * Crash on shutdown (Null pointer dereference)

turtleli commented:
Turns out it wasn't a null pointer dereference. I didn't check if the RTTI settings were consistent - it was on for wxWidgets 3.0 projects but off for every other project. So I turned it off and it doesn't crash on shutdown.

There's one bug that I can reliably reproduce that isn't present with 2.8 on Windows, though it shouldn't be encountered under normal circumstances.
Steps to reproduce:
1. Have the first time wizard disabled and delete the inis.
2. Run pcsx2, choose plugins and game but not BIOS.
3. Choose either Boot CDVD option or choose Run through the Debugger.

With a 2.8 build it tells you your BIOS isn't configured and tells you to configure it. With 3.0, the same messagebox pops up, promptly hides itself and gives focus to the next topmost non PCSX2 window, and you have to use Task Manager to kill PCSX2.

It might be due to a behaviour change in wx3.0. I think there's actually a hard to encounter bug with a Linux wx3.0 build (tried on both Fedora GCC 5.1.1 and Slackware GCC 4.8.2, I've compared it with a wx2.8 build) where the problem code lies in the same area, though with less dire effects on a release build.

Steps to reproduce the Linux wx3.0 bug (GTK2/3 doesn't matter)
1. Have the first time wizard disabled and delete the inis.
2. Run pcsx2. Choose any option to close the messagebox. If you use a debug build it will assert and terminate. For release builds it will fail to open the plugin selector panel if you select ok, though it still works fine if you open it from the menu.

gregory38 commented:
Note: wx2.8 version was kinds of tuned for PCSX2. Maybe it could explain the difference. You can spot the difference if you search "PCSX2" string in 3rdparty code.

ramapcsx2 commented:
I remember the BIOS missing message box was a straight hack for an annoying situation Jake had in that situation.
Maybe that hack was required because of insufficiencies in wx 2.8?
It's worth a try to re-think / rebuild that part.

turtleli commented:
That seems to be it. The emulation thread and the messagebox are interfering with each other. As a quick test I threw in a CoreThread.Suspend() before the messagebox appears and it doesn't lock up anymore. So I just need to do a proper fix, do a lot more testing and perhaps I'll be done.

Perhaps the best thing to do would be to check that the plugin and bios are properly configured and if not, disable the boot cdvd and debugger run options. But I'll leave rethinking/rebuilding for later/much later/someone else/preferably someone else. Dealing with both 2.8 and 3.0 is a bit much and requires far too much testing.

The modifications to the 2.8 source seems to be for memory allocation and disabling some debugging.

turtleli commented:
I've reworked how missing BIOS is handled. This:
 * fixes a lockup on a Windows wxWidgets 3.0 build and fixes a crash on Linux wxWidgets 2.8 and  wx3.0 builds;
 * fixes how the message box looks; and
 * makes the BiosSelectorDialog unused, so I've removed it and adjusted the VS and cmake file.

I've also modified a few VS2012 files that I forgot to do before (merged into earlier commits).

That should be it for changes unless someone finds that I've broken something. Tested on Windows wx2.8 and wx3.0, Linux wx2.8 and wx3.0(GTK3)

Note: The Game Database Editor has increased in height for some reason on a Windows wxWidgets 3.0 build. It still works though, so I won't be making changes to it.

gregory38 commented:
Honestly, good jobs. You're ours wx savior.

avih commented on pcsx2/gui/AppMain.cpp:
```diff
@@ -236,8 +291,9 @@ void Pcsx2App::PadKeyDispatch( const keyEvent& ev )
+//returns 0 for normal keys and a WXK_* value for special keys
 #ifdef __WXMSW__
-	const int vkey = wxCharCodeMSWToWX( ev.key );	//returns 0 if plain ascii value or a WXK_... (<=32 or >=300) if a special key
+	const int vkey = TranslateVKToWXK(ev.key);
```
Why did you add your own translation function? It seems to me that according to http://trac.wxwidgets.org/ticket/12508 , the only difference between the new `VKToWX` to the old `wxCharCodeMSWToWX` is that the new one doesn't return 0, i.e. it returns either an ascii code or the special code, at which case I'd _think_ it should work as is by calling `VKToWX` on wx3.x, since the rest of the code does `m_kevt.m_keyCode = vkey? vkey : ev.key;` which should work either for the old wx call or the new one.

Am I wrong?

turtleli replied:
I wanted to have both Windows and Linux behave identically at that point. It seems you're most likely correct that I could have just used the newer function and it wouldn't have made a difference.

avih replied:
Well, it seems to me that for linux it uses `TranslateGDKtoWXK` anyway, and your new function `TranslateVKToWXK` only translates `VK_` which AFAIK is windows only... am I missing something?

turtleli replied:
I started from `TranslateGDKtoWXK` (not a wx function) and then converted the GDK_ keys to the VK_ equivalents.

avih replied:
I don't yet have setup to build with wx3, but could you please try to use `VKToWX` for wx3 only and see if it works? also try to verify if it fixes #674 .

In general, unless there's a good reason, I think it's better not to rewrite existing library functions if they work well.

turtleli replied:
Alright, will do. If it works, I'll revert commit 760de9915bb54d1243f8afc95f7d630ca3eb54ec then apply the new commit for the PR.

avih replied:
Appreciated.

turtleli replied:
Hmm. `VKToWX` can't be used. It's an internal function for wx3.0, available only in a private header, and grep didn't return a non-private header I could use for it.

avih replied:
OK, thanks for checking it out. Maybe we'll try to look into it again at a later time.

Don't forget #674 ;)
